Tests for tracking known transactions

rvl opened the pull request:
Relates to issue #90 

### Overview

This is a start for tests of the transaction tracking part of `applyBlock`.

denisshevchenko commented on test/unit/Golden/golden/apiV1Types/json/Timestamp:
```diff
@@ -0,0 +1 @@
+"2018-12-11T08:09:10"
```
> (would be nice if the commit history reflect this wink)

Sorry, I rebased (squash) it... :-)

rvl committed:
Remove unneeded GHC extension

rvl committed:
Add a couple properties for transaction tracking

KtorZ commented on test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -129,6 +148,50 @@ prop_applyBlockBasic s =
+prop_applyBlockTxHistoryIncoming :: WalletState -> Property
+prop_applyBlockTxHistoryIncoming s = conjoin (map prop wallets)
+  where
+    wallets = scanl (flip applyBlock) (initWallet s) blockchain
```
Is there a particular reason to check all intermediate state and not only the last one :thinking:  ?

KtorZ commented on test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -129,6 +148,50 @@ prop_applyBlockBasic s =
+        inTxs = filter isIncoming $ Set.toList $ getTxHistory wallet
+        isIncoming (_, m) = direction m == Incoming
+        txOuts = map (Set.fromList . map address . outputs . fst) inTxs
+        overlaps a b = not (Set.disjoint a b)
```
It'll be nice to add a coverage checks that makes sure that `addrs` and `ourAddresses` aren't just empty all the time.

KtorZ commented on test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -85,6 +96,14 @@ spec = do
+        it "Incoming transactions have output addresses that belong to the wallet"
+            (property prop_applyBlockTxHistoryIncoming)
+        it "Any transction involving our addresses must be incoming"
+            (property prop_applyBlockTxHistoryOurs)
```
I don't get this second property :thinking: ? What about outgoing transactions with change?

KtorZ committed:
backup

KtorZ committed:
remove commented-out property & adjust metadata property with latest master

KtorZ committed:
add more previous blocks to the 'fake' blockchain from the primitive model tests

This allows for having the original transactions that are referred to in some other blocks, and therefore, counts for outgoing transactions
as well during testing

KtorZ committed:
add property test about wallet current tip

KtorZ committed:
remove 'getTxHistory' accessor: in the long run, we may not even need to keep the tx history stored in the wallet state

KtorZ reviewed (approved):
reviewed + extended test data to cover more cases / branches from the model.
